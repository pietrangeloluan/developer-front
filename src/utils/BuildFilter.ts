export default function (params: any) {
  let filter = {}
  for (const key in params) {
    if (Object.hasOwnProperty.call(params, key)) {
      const parameter = params[key]
      if (parameter) {
        filter = { ...filter, [key]: parameter }
      }
    }
  }
  return filter
}
