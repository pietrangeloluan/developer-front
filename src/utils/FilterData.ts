export default function (obj, filter) {
  const filtered = obj.filter(item => {
    for (const key in filter) {
      if (item[key] === undefined || !filter[key].includes(item[key])) {
        return false
      }
    }
    return true
  })

  return filtered
}
