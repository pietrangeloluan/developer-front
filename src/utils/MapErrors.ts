export function mapErrors(errors, context) {
  const errorKeys = Object.keys(errors)
  const regex = new RegExp(`${context}\\[\\d+\\]`, 'g')
  const contextErrors = errorKeys.filter(err => err.match(regex))

  if (!contextErrors.length) return [{}]

  const updated = []

  contextErrors.forEach(err => {
    const found = [...err.matchAll(/\[(\d+)\]\.(\w+)/g)]

    if (found.length) {
      found[0].shift()
      const errIndex = found[0][0]
      const errField = found[0][1]

      updated[errIndex] = {
        ...updated[errIndex],
        [errField]: errors[`${context}[${errIndex}].${errField}`]
      }
    }
  })

  return updated
}
