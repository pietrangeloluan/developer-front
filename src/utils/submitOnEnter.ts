import { KeyboardEvent } from 'react'

export function submitOnEnter(fn: Function, args: any[] = []) {
  return function (e: KeyboardEvent) {
    if (e.key === 'Enter') fn.apply(args)
  }
}
