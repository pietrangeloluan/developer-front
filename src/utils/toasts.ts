import { toast } from 'react-toastify'

export const toasts = {
  created() {
    return toast.success('Dados cadastrados com sucesso.', {
      autoClose: 5 * 1000
    })
  },
  deleted() {
    return toast.success('Dados exclu√≠dos com sucesso.', {
      autoClose: 5 * 1000
    })
  },
  edited() {
    return toast.success('Dados editados com sucesso.', { autoClose: 5 * 1000 })
  },
  generalFail({ code = 'UNKNOWN_ERROR' } = {}) {
    return toast.error(
      'Ocorreu uma falha inesperada. Por favor, aguarde alguns minutos de tente novamente.',
      { autoClose: 5 * 1000, toastId: code }
    )
  }
}
