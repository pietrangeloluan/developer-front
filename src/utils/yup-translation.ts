/* eslint-disable no-template-curly-in-string */
import { setLocale } from 'yup'

const translation = {
  array: {
    max: '${path} deve conter no máximo ${max} itens',
    min: '${path} deve conter no mínimo ${min} itens'
  },
  date: {
    max: '${path} deve ser anterior à ${max}',
    min: '${path} deve ser mais recente que ${min}',
    required: '${path} é obrigatório'
  },
  mixed: {
    default: '${path} é inválido',
    notOneOf: '${path} não pode ser igual à um destes valores: ${values}',
    oneOf: '${path} deve ser igual à um dos seguintes valores: ${values}',
    required: '${path} é obrigatório'
  },
  number: {
    integer: '${path} deve ser inteiro',
    lessThan: '${path} deve ser menor à ${less}',
    max: '${path} deve ser no máximo ${max}',
    min: '${path} deve ser no mínimo ${min}',
    moreThan: '${path} deve ser maior à ${more}',
    negative: '${path} deve ser negativo',
    notEqual: '${path} não pode ser igual à ${notEqual}',
    positive: '${path} deve ser positivo',
    required: '${path} é obrigatório'
  },
  string: {
    email: '${path} não contém um e-mail válido',
    length: '${path} deve conter exatamente ${length} caracteres',
    lowercase: '${path} deve ser maiúsculo',
    max: '${path} deve conter no máximo ${max} caracteres',
    min: '${path} deve conter pelo menos ${min} caracteres',
    trim: '${path} não deve conter espaços em branco em início ou final.',
    uppercase: '${path} deve ser minúsculo',
    url: '${path} não contém uma URL válida'
  }
}

setLocale(translation)
