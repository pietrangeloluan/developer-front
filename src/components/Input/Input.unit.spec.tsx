import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import faker from 'faker'

import { renderWithThemeProvider } from '__tests__/utils'

import { Input } from '.'

describe('Input', () => {
  it('should render the component', () => {
    renderWithThemeProvider(Input, { 'data-testid': 'input' })

    const found = screen.getByTestId('input')
    expect(found.firstChild).toBeInTheDocument()
  })

  it('should renders with default value', () => {
    const label = faker.lorem.word()
    const defaultValue = faker.lorem.words()

    renderWithThemeProvider(Input, {
      label,
      defaultValue
    })

    const found = screen.getByDisplayValue(defaultValue)
    expect(found).toBeInTheDocument()
  })

  it('should renders with Label', () => {
    const label = faker.lorem.word()
    const id = faker.lorem.word()
    renderWithThemeProvider(Input, { label, id })

    const found = screen.getByLabelText(label)
    expect(found).toBeInTheDocument()
  })

  it('should renders with placeholder', () => {
    const placeholder = faker.lorem.words()
    renderWithThemeProvider(Input, { placeholder })

    const found = screen.getByPlaceholderText(placeholder)
    expect(found).toBeInTheDocument()
  })

  it('should have the variant outlined as default and cannot be changed', () => {
    const { container } = renderWithThemeProvider(Input)

    const test = container.children.item(0).children.item(0).children.item(0)

    expect(test.firstChild).toHaveClass('MuiOutlinedInput-root')
  })

  it('should has light border as default', () => {
    const { container } = renderWithThemeProvider(Input)

    const test = container.children.item(0).children.item(0)

    expect(test.firstChild).toHaveAttribute('border', 'light')
  })

  it('should has another border when receives it', () => {
    const border = 'dark'
    const { container } = renderWithThemeProvider(Input, { border })

    const test = container.children.item(0).children.item(0)

    expect(test.firstChild).toHaveAttribute('border', border)
  })

  it('should has medium padding as default', () => {
    const { container } = renderWithThemeProvider(Input)
    const test = container.children.item(0).children.item(0)
    expect(test.firstChild).toHaveAttribute('padding', 'medium')
  })

  it('should has another padding when receives it', () => {
    const padding = 'large'
    const { container } = renderWithThemeProvider(Input, { padding })
    const test = container.children.item(0).children.item(0)
    expect(test.firstChild).toHaveAttribute('padding', padding)
  })

  it('should changes its value when typing', async () => {
    const onInput = jest.fn()
    const label = faker.lorem.word()
    const id = faker.lorem.word()

    renderWithThemeProvider(Input, { label, id, onInput })

    const input = screen.getByRole('textbox')
    const text = faker.lorem.words()
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('should is accessible by tab', () => {
    const label = faker.lorem.word()
    const id = faker.lorem.word()

    renderWithThemeProvider(Input, { label, id })

    const input = screen.getByLabelText(label)
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('should renders with icon left', () => {
    const label = faker.lorem.word()
    const iconLeft = 'input-user'
    renderWithThemeProvider(Input, {
      label,
      iconLeft
    })

    const found = screen.getByText(iconLeft)
    expect(found).toBeInTheDocument()
  })

  it('should renders with icon right', () => {
    const label = faker.lorem.word()
    const iconRight = 'input-user'
    renderWithThemeProvider(Input, {
      label,
      iconRight
    })

    const found = screen.getByText(iconRight)
    expect(found).toBeInTheDocument()
  })

  it('should renders with errors', () => {
    const error = true
    const helperText = faker.lorem.words()
    renderWithThemeProvider(Input, { error, helperText })

    const found = screen.getByText(helperText)
    expect(found).toBeInTheDocument()
  })

  it('should match the snapshot', () => {
    const { container } = renderWithThemeProvider(Input, {
      value: 'Hello World!'
    })
    expect(container).toMatchSnapshot()
  })
})
