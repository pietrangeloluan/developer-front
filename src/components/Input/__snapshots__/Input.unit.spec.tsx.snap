// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match the snapshot 1`] = `
.c0 {
  margin: 0.8rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 .MuiInputBase-root {
  font-size: 1.4rem;
  line-height: 2.4rem;
  color: #464a53;
  padding: 0 1.6rem;
  padding-left: 1.6rem;
  padding-right: 1.6rem;
}

.c2 .MuiInputBase-root .MuiIconButton-root {
  position: absolute;
  right: 4px;
}

.c2 .MuiInputLabel-root {
  font-size: 1.4rem;
  -webkit-transform: translate( 1.6rem, 1.3rem ) scale(1);
  -ms-transform: translate( 1.6rem, 1.3rem ) scale(1);
  transform: translate( 1.6rem, 1.3rem ) scale(1);
}

.c2 .MuiInputLabel-outlined.MuiInputLabel-shrink {
  -webkit-transform: translate(14px,-6px) scale(0.71428);
  -ms-transform: translate(14px,-6px) scale(0.71428);
  transform: translate(14px,-6px) scale(0.71428);
}

.c2 .MuiOutlinedInput-input {
  padding: 1.17rem 0;
}

.c2 .MuiOutlinedInput-input.MuiOutlinedInput-inputMultiline {
  padding: 1.17rem 0;
}

.c2 label.Mui-focused {
  color: #408ae4;
}

.c2 .MuiOutlinedInput-root fieldset {
  opacity: 1;
  -webkit-transition: border-color 400ms;
  transition: border-color 400ms;
  will-change: border-color;
}

.c2 .MuiOutlinedInput-root:hover fieldset {
  border-color: #0f0f0f;
}

.c2 .MuiOutlinedInput-root.Mui-focused fieldset {
  border-color: #408ae4;
}

.c2 .MuiOutlinedInput-root.Mui-error .MuiOutlinedInput-notchedOutline {
  border-color: #de4f4f;
}

.c2 .MuiFormLabel-root.Mui-error {
  color: #de4f4f;
}

.c2 .MuiFormHelperText-root.Mui-error {
  color: #de4f4f;
  font-size: 1rem;
  font-weight: 600;
  padding-left: 0.4rem;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        border="light"
        class="MuiFormControl-root MuiTextField-root c2"
        icon="[object Object]"
        padding="medium"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            type="text"
            value="Hello World!"
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-57 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-58"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
